### PR TITLE
Define entity-attribute model for ClojureDocs data layer

### PR DESCRIPTION
**Pull request by lambduhh**
## Context

Addresses first part of #43. The ClojureDocs data layer had no explicit schema — entity shapes were implicit in `search.clj` (JVM-heap-derived) and in scattered MongoDB writes. This PR makes the model explicit, machine-readable, and **verified against the running system**, as groundwork for the multi-library/multi-version work in the 2026 vision.

This branch has since been synced with `master` to adopt the repo-wide OKF + RDF document-metadata convention (#71); its docs are migrated to YAML frontmatter and pass the metadata validator (see "Also in this PR" below).

## Problem

- No single source of truth for which entities/attributes exist, their types, or their coverage.
- The boundary between JVM-heap-derived entities (Library, Namespace, Var) and MongoDB-persisted ones (Example, Note, SeeAlso, …) was undocumented.
- Latent bugs were invisible because nothing pinned the data shapes.

## Solution

- **Source of truth — [`docs/entity-attribute-model.edn`](docs/entity-attribute-model.edn)** (supersedes the earlier CSV): 11 entities with per-attribute `:type`, `:required?`, `:coverage`, and a `:status` enum (`:exists` / `:nil` / `:vestigial` / `:absent` / `:planned` / `:gap`). Schema v1.1.0.
- **Enforced by tests — [`test/clojuredocs/entity_model_test.clj`](test/clojuredocs/entity_model_test.clj)** (25 tests, ~12.5k assertions): the JVM-heap entities and DialectCompat are checked against the running system on every run — the reliability ratchet, with REPL findings hardened into guards. MongoDB cardinalities remain snapshot-derived and are **not yet test-guarded** (#66).
- **Supporting docs** — [RFC](docs/rfcs/entity-model-rfc.md) (design rationale), [decisions log](docs/decisions.md) (format/diagram choices), [errata](docs/errata.md) (errors found and corrected during extraction + verification). A Miro visual is an eventual goal — data first, per review with Alex — and is not yet produced.
- **Bugs surfaced & filed** — #53, #54, #55, #56 (during extraction); #66–#70 (during the verification pass: Mongo test gap, unreachable `:type "special-form"`, version-locking, dialect-compat count, vestigial indexing).

## Also in this PR — OKF metadata sync

- Merged `master` to pick up the OKF + RDF document-metadata convention (#71): CLAUDE.md spec, `docs/context.jsonld`, `docs/index.md`.
- Migrated this branch's docs from the old `> **Document metadata**` blockquote to OKF YAML frontmatter — `entity-attribute-model.md`, `errata.md`, `rfcs/entity-model-rfc.md`, `diagrams/dataflows.md`, and the research-review run-1 log. Provenance/maturity values preserved; disclaimers and the scope note moved into the body.
- Brought in `tools/validate_metadata.clj` + `docs/metadata-schema.edn` (from the enforcement branch, draft #72) so the branch self-validates: **21/21 docs pass**. The enforcement *wiring* (pre-commit hook, `lein test`, CI) lands with #72; a later re-merge of `master` de-duplicates these two files.

## Review maturity

L4 — human-endorsed (Jordan Miller) for the JVM-heap entities (REPL-verified + test-guarded). MongoDB cardinalities held at L2 (snapshot-derived, #66).

**Scope — present-state only, so far.** #43 is to map the full data model: what exists today **and** what the 2026 vision will require us to build. This PR delivers the **verified present-state** (`:status :exists`, checked against the running system). The **envisioned future-state** — combing the vision to enumerate entities/attributes still to be created (`:status :gap` / `:planned`) — is the pending second half of #43 and is not in this PR (beyond a few `:gap` attrs on `:dialect-compat`). So "verified"/L4 reflects the present-state half, not completeness against the vision.

<details>
<summary>Father Watson Questions</summary>

**What do we know?** The JVM-heap entity shapes are exact — REPL-verified and locked to Clojure 1.12.4 by the test suite. The model's structure (entities, embedded sub-schemas, status enum) matches `api/common.clj` and `search.clj`.

**What do we need to know?** The MongoDB cardinalities (examples 2671, users 4902, …) come from one snapshot via `dev/sidecar_repl.clj` — they will drift and aren't guarded (#66). Whether `:type "special-form"` should be reachable end-to-end (#67) is a product decision.

**Where are we?** The model is the verified, test-guarded source of truth for the in-process entities; errata records what was wrong and corrected. Docs now follow the OKF metadata convention and pass the validator.

**Where are we going?** Close the Mongo verification gap (#66), then this can move to L4. The `:gap` entities are the substrate for multi-library support.

</details>

🧬 Recommended order for a human reviewer to consume:
- Start with the decision log to grep context
- Read/review RFC/entity-model
- see entity-model.edn

### CODE REVIEW

File: docs/decisions.md [hunk 1]
```diff
@@ -1,9 +1,393 @@
...
+
+### Status
+Decided
+
+### Context
+- Sandra reviewed PR #57 and discovered Claude produced two separate ER diagrams instead of the single unified diagram specified in the 2026-06-01 decision entry.
+- The entity-model.md header says "Solid lines = exists today. Dashed lines = required by vision" but the diagrams don't deliver this — they are two disconnected `erDiagram` blocks.
+- No formal legend or key exists for reading the diagram.
+- No order-of-magnitude grounding: the diagram doesn't communicate that there is 1 Library (hardcoded singleton), ~30 Namespaces, ~700 Vars, and thousands of Examples.
+- Sierra's ["Developing the Language of the Domain"](https://www.cognitect.com/blog/2017/4/6/developing-the-language-of-the-domain) (Cognitect, 2017) reinforces that models need facets and ground truth to be useful — a diagram without a legend or sense of scale fails to communicate the domain.
```
Comment (lambdasierra) [inline docs/decisions.md]: I appreciate the shout-out, but rest of the sentence doesn't make sense. The two clauses on either side of the em-dash are unrelated. My blog post doesn't say anything about diagram legends or scale.

File: docs/decisions.md [hunk 2]
```diff
@@ -1,9 +1,393 @@
...
+
+## 2026-06-05 — Correct entity model diagram manually in Miro
+
+### Status
+Decided
+
+### Context
+- The AI-generated ER diagram in PR #57 deviated from the stated design (two diagrams instead of one, no legend, no magnitude).
+- Miro is the team's standard diagramming tool. Mermaid ER diagrams have rendering limitations (no dashed lines, no annotations) that constrain what can be expressed in-repo.
+- Jordan's pedagogical framework is constructivism: learning happens through the act of constructing and correcting mental models, not through passively receiving correct output.
```
Comment (lambdasierra) [inline docs/decisions.md:539]: Nice! I really like this lens. It's a justification for **doing the work**, instead of **letting the AI do it.**

File: docs/decisions.md [hunk 3]
```diff
@@ -1,9 +1,393 @@
...
+- Sandra recommended correcting the diagram by hand as the most effective way to internalize the domain model.
```
Comment (lambdasierra) [inline docs/decisions.md]: Did I? I don't remember saying this. 😆 

What I do remember saying is that it's hard to convey meaning through spatial relationships when the layout of the diagram is algorithmically-generated.

Humans infer meaning from spatial relationships: left-to-right, top-to-bottom, distance. When a person arranges a diagram by hand, we intuitively express our mental model: _X_ is "above" _Y_; _Z_ and _Q_ are "close". This is an extra layer of information that generated diagrams usually lack, and it's hard to get a generator to follow all of the intuitive relationships we hold in our heads.

On the other hand, generated diagrams have the advantage of showing **only** the relationships that are actually expressed in the model, without any hidden assumptions.

File: docs/decisions.md [hunk 4]
```diff
@@ -1,9 +1,393 @@
...
+### Context
+- Mermaid's `erDiagram` parser treats hyphens as operators and rejects `{}`, `|`, `()`, and unicode characters in attribute descriptions. Clojure uses kebab-case pervasively, making every attribute name a syntax error.
+- Fixing the diagrams required mangling all attribute names to underscores (`library-url` → `library_url`) and stripping descriptive annotations — losing fidelity to the actual domain.
+- The [2026-06-05 decision to correct the diagram in Miro](#2026-06-05--correct-entity-model-diagram-manually-in-miro) already established Miro as the canonical visual.
+- Sierra's ["Developing the Language of the Domain"](https://www.cognitect.com/blog/2017/4/6/developing-the-language-of-the-domain) describes this exact pattern: the domain model is data, visualizations are derived views.
+- The repo already has precedent for EDN as a data format: `dialect-compat.edn`.
+
+### Decision
+- Remove Mermaid ER diagrams from [entity-attribute-model.md](docs/entity-attribute-model.md).
+- Use EDN as the machine-readable source of truth for the entity model, with a `:status` field (`:exists` / `:gap` / `:vestigial`) on each entity and attribute.
```
Comment (lambdasierra) [inline docs/decisions.md:475]: 👍🏻

File: docs/decisions.md [hunk 5]
```diff
@@ -1,9 +1,393 @@
...
+### Rationale
+- Keeps production source files clean — no exploratory code in diffs or PRs.
+- Reduces AI context window cost — AI tools reading source files don't ingest investigation noise.
+- The `comment` block pattern makes forms individually evaluable at the REPL without side effects on load.
+- Scratchpad is committed so investigation history is visible and reproducible.
+
+### Alternatives Considered
+- Rich comment forms in source files — conventional in Clojure, but mixes investigation with production code and inflates diffs.
+- Ephemeral REPL history only — loses the investigation record; not reproducible by reviewers.
+- Separate `.repl` files (transcriptor-style) — appropriate for verification tests but too structured for exploratory investigation.
```
Comment (lambdasierra) [inline docs/decisions.md:454]: I agree 💯 with keeping exploratory code out of production source files.

The weakness of the "sidecar" approach is that the code isn't exercised regularly, so it quickly becomes out-of-date. It's also tedious to run by manually stepping through in an editor.

If you want to keep your 'sidecar' files, do it exclusively as a historical reference, linked to a specific date or version of the code. Anything that needs to be re-tested or read — either by AI or humans — should be executable.

[Transcriptor](https://github.com/cognitect-labs/transcriptor)-style REPL files, or just plain old unit tests, are reproducible and automatable.

File: docs/decisions.md [hunk 6]
```diff
@@ -1,9 +1,393 @@
...
+- The [2026-04-28 RCF decision](#2026-04-28--verify-dialect-compat-via-rich-comment-blocks-not-unit-tests) chose rich comment blocks over unit tests because "adding a unit test to a project with zero real test coverage sets a convention nobody is following yet."
+- That tradeoff shifted: the model produces dozens of exact, REPL-verified facts (var count 1,572; per-key coverage; the special-form/type behavior; embedded-doc uniformity) worth locking against regression and Clojure-version drift.
+
+### Decision
+- Add [`test/clojuredocs/entity_model_test.clj`](../test/clojuredocs/entity_model_test.clj) (25 tests, ~12.5k assertions). It loads the EDN and checks the JVM-heap entities (Library, Namespace, Var) and DialectCompat against the running system on every run. MongoDB entities are intentionally excluded — they need a DB connection and are snapshot-derived (tracked in [#66](https://github.com/nubank/clojuredocs/issues/66)).
+
+### Rationale
+- Hardens REPL findings into enforcement — the final ratchet step. A regression or a Clojure bump now fails a test instead of silently rotting the doc.
```
Comment (lambdasierra) [inline docs/decisions.md:341]: 👍🏻

File: docs/decisions.md [hunk 7]
```diff
@@ -1,9 +1,393 @@
...
+### Status
+Decided
+
+### Context
+- The [2026-06-05 entries](#2026-06-05--replace-mermaid-er-diagrams-with-edn-schema-and-miro-visual) named Miro the "canonical visual" for the entity model, with a PDF export to be checked into the repo.
+- In practice no Miro diagram or PDF was ever produced or committed (`git ls-files` shows no PDF), yet the docs described it as the current canonical representation.
+- Reviewing with Alex: he can review the data (the EDN) directly — a diagram is not a prerequisite for review.
+
+### Decision
+- Prioritize sound, verified data (the EDN) as the first deliverable and the review artifact. Treat the Miro visual as the eventual/ideal representation, deferred until the data is solid — not a blocker for review.
```
Comment (lambdasierra) [inline docs/decisions.md:303]: Makes sense, although it might be harder to review without a visual representation.

This is a challenge with every diagramming method I have tried: A diagram is only a **partial** representation of the model at a **single** point in time. It gets stale quickly.

Some architecture modeling tools (such as [ArchiMate](https://nubank.atlassian.net/wiki/spaces/~5fb5637dcbead50069e9aa4a/pages/264398801219/ArchiMate+Archi+and+ADOIT)) support visual diagrams that are still linked to the underlying data model, at the cost of a much more complex model and tool lock-in. I've been trying to figure out how to apply this in a more Clojure-y way, but without success so far.

**Review by lambdasierra — COMMENTED** [review]
Just commenting on the decision log. Still need to review the actual entity-relationship model.

File: docs/entity-attribute-model.edn [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+  {:login          {:type :string :required? true :description "Username (GitHub or legacy)"}
+   :account-source {:type :string :required? true :description "Auth provider: \"github\" or \"clojuredocs\""}
+   :avatar-url     {:type :string :required? true :description "Avatar image URL"}}}
+
+ :entities
+ {:library
+  {:source "jvm-heap"
+   :cardinality 1
+   :description "Singleton — the Clojure standard library. Hardcoded in search.clj."
+   :attrs
```
Comment (puredanger) [inline docs/entity-attribute-model.edn:53]: We should have library attributes for:

1. deps.edn style lib name (`groupId/artifactId`)
2. deps.edn style coordinate (this is a broader concept than version, although there is an api in tools.deps to obtain a printable "version" from the coordinate, useful for display). This is generic, so works for both maven and git (and local and whatever else we add in the future) deps.
3. description?  once we have a list of libraries we will want a 1 line description - this could be pulled from the git project.
4. namespaces to include in docs - every library has a subset that is the "public" API to document (a subset of the full list). Maybe that's covered by :namespaces if that some human is involved in deciding what's in it.
5. lots of other potential metadata, but maybe license is the next most salient. SPDX is a standard set of codes for this, which of course, no one really uses well

By having a flat library record that includes version we are implicitly allowing multiple versions per lib. A big question - do we support every version of a lib? only newest? only ones some human decided to include? we should have a concrete answer to that.

File: docs/entity-attribute-model.edn [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+                              :evidence "sidecar_repl.clj — (dissoc search/clojure-lib :vars :namespaces)"}}
+    :source-base-url {:type :string :required? true
+                      :description "GitHub blob URL prefix for source links"
+                      :status {:state :exists :verified "2026-06-09"
+                               :evidence "sidecar_repl.clj — (dissoc search/clojure-lib :vars :namespaces)"}}
+    :gh-tag-url     {:type :string :required? true
+                     :description "GitHub tag URL for the release"
+                     :status {:state :exists :verified "2026-06-09"
+                              :evidence "sidecar_repl.clj — (dissoc search/clojure-lib :vars :namespaces)"}}
+    :namespaces     {:type :vector :required? true
```
Comment (puredanger) [inline docs/entity-attribute-model.edn]: these vector types are really implicitly refs (in Datomic speak) and it would be helpful to know what entity type they refer to (obvious here of course)

File: docs/entity-attribute-model.edn [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+                            :evidence "sidecar_repl.clj — var key universe scan"}}
+    :doc          {:type :string  :required? false :coverage "1293/1572"
+                   :description "Docstring"
+                   :status {:state :exists :verified "2026-06-09"
+                            :evidence "sidecar_repl.clj — var key universe scan"}}
+    :added        {:type :string  :required? false :coverage "907/1572"
+                   :description "Version when this var was added"
+                   :status {:state :exists :verified "2026-06-09"
+                            :evidence "sidecar_repl.clj — var key universe scan"}}
+    :static       {:type :boolean :required? false :coverage "319/1572"
```
Comment (puredanger) [inline docs/entity-attribute-model.edn:146]: this no longer has meaning, can be dropped

File: docs/entity-attribute-model.edn [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+                   :description "Static compilation hint (legacy)"
+                   :status {:state :exists :verified "2026-06-09"
+                            :evidence "sidecar_repl.clj — var key universe scan"}}
+    :macro        {:type :boolean :required? false :coverage "190/1572"
```
Comment (puredanger) [inline docs/entity-attribute-model.edn:150]: rather than having attrs for :macro, :special-tag, etc should better have some kind of "var-type" attribute

File: docs/entity-attribute-model.edn [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+                            :evidence "sidecar_repl.clj — var key universe scan"}}
+    :tag          {:type :string  :required? false :coverage "122/1572"
+                   :description "Return type hint (stringified from class)"
+                   :status {:state :exists :verified "2026-06-09"
+                            :evidence "sidecar_repl.clj — var key universe scan"}}
+    :dynamic      {:type :boolean :required? false :coverage "56/1572"
+                   :description "True if this var is dynamic (e.g. *out*)"
+                   :status {:state :exists :verified "2026-06-09"
+                            :evidence "sidecar_repl.clj — var key universe scan"}}
+    :skip-wiki    {:type :boolean :required? false :coverage "31/1572"
```
Comment (puredanger) [inline docs/entity-attribute-model.edn:162]: more generically, maybe we should have some kind of ":impl" flag attribute rather than the weird :skip-wiki.and :no-doc stuff.

File: docs/entity-attribute-model.edn [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+    :created-at {:type :integer :required? true
+                 :description "Unix epoch millis when created"
+                 :status {:state :exists :verified "2026-06-09"
+                          :evidence "sidecar_repl.clj — notes key universe scan"}}
+    :updated-at {:type :integer :required? true
+                 :description "Unix epoch millis when last updated"
+                 :status {:state :exists :verified "2026-06-09"
+                          :evidence "sidecar_repl.clj — notes key universe scan"}}}}
+
+  :see-also
```
Comment (puredanger) [inline docs/entity-attribute-model.edn:284]: All these user-created things share a lot of commonality and could be the same "entity" type in Datomic with shared (some optional) attributes. In Datomic, there are no entity types (although that's helpful to understandd your model), just attributes.

File: docs/entity-attribute-model.edn [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+    :name         {:type :string :required? true
+                   :description "Var name"
+                   :status {:state :exists :verified "2026-06-09"
+                            :evidence "sidecar_repl.clj — legacy-var-redirects key universe scan"}}
+    :library-url  {:type :string :required? true
+                   :description "Library GitHub URL"
+                   :status {:state :exists :verified "2026-06-09"
+                            :evidence "sidecar_repl.clj — legacy-var-redirects key universe scan"}}}}
+
+  :dialect-compat
```
Comment (puredanger) [inline docs/entity-attribute-model.edn:361]: this will probably want a richer schema over time. we just have flags now basically, but we really want/need a place for dialect compatibility notes, like "ClojureScript supports case, but it evals the cases where Clojure does not". Stuff like that is the gold and we have no place for it tied to dialects.

**Review by puredanger — COMMENTED** [review]
I glanced through some of the stuff but really only read the high level of the edn schema and left some notes there. 

We have lots of prior art for representing Clojure code as namespace and var attributes, and we really want to persist this model, not (just) load it in memory for search - that lets us do way more powerful things. Existing data  formats for this exist - [autodoc](https://github.com/tomfaulhaber/autodoc-collect#problem) (Clojure api docs), [clj-kondo analysis data](https://github.com/clj-kondo/clj-kondo/blob/master/analysis/README.md), codeq, I'm sure there are others. I'd be inclined to lean hard on clj-kondo analysis as a very active living project. We might want a lot *less* in the db (really just the skeleton) and more in analysis files we load from a store like s3. I don't know - figuring out what makes sense to put in db / file store / memory is an interesting part of the problem (and clearly the current split has issues in the existing site by requiring stuff on the cp to load it into search).

File: docs/entity-attribute-model.edn (hunk 12 above)
Comment (lambdasierra) [reply docs/entity-attribute-model.edn:162]: I'm assuming that these flags were never documented anywhere. If that's true (@puredanger ?) then I agree, they should probably be isolated in an "implementation detail" entity.

File: docs/entity-attribute-model.edn (hunk 13 above)
Comment (lambdasierra) [reply docs/entity-attribute-model.edn:284]: In theory, yes. It's easy to carry this too far, though. If everything is the same "type" of thing, then it's a lot more work to answer straightforward questions like "Find all the Notes where …". So in a future Datomic schema, I would recommend different "ID" attributes for each "type" of thing, like `:note/id`, `:example/id`, and so on.

On the other hand, reusing common attributes makes it **easier** to do queries like "Find all the user-generated things that reference Var X." This is where you could shared attributes, something like `:commentary/var` for "this entity refers to this Var."

File: docs/entity-attribute-model.edn (hunk 14 above)
Comment (lambdasierra) [reply docs/entity-attribute-model.edn:361]: Agreed, this will be one of the most active opportunities of enhancement.

File: docs/entity-attribute-model.edn [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+                          :evidence "sidecar_repl.clj — examples key universe scan"}}
+    :body       {:type :string :required? true
+                 :description "The example source code"
+                 :status {:state :exists :verified "2026-06-09"
+                          :evidence "sidecar_repl.clj — examples key universe scan"}}
+    :author     {:type :ref :schema :embedded-user :required? true
+                 :description "User who created the example"
+                 :status {:state :exists :verified "2026-06-09"
+                          :evidence "sidecar_repl.clj — examples key universe scan"}}
+    :created-at {:type :integer :required? true
```
Comment (lambdasierra) [inline docs/entity-attribute-model.edn:209]: In Datomic, **in theory**, you can derive created-at and updated-at from Datomic history. But in practice, it's not trivial. 

Even though they are technically "redundant" in Datomic, created-at and update-at fields can be useful as a cache if you need to display this information in a UI.

**Review by lambdasierra — COMMENTED** [review]
This is my first pass review of the entity schema.

File: docs/entity-attribute-model.edn [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+                     :evidence "entity_model_test.clj — namespace-universal-keys"}}
+    :added {:type :string :required? false :coverage "2/38"
+            :description "Version when namespace was added. Only clojure.pprint and clojure.reflect."
+            :status {:state :exists :verified "2026-06-09"
+                     :evidence "sidecar_repl.clj — namespace key coverage"}}}}
+
+  :var
+  {:source "jvm-heap"
+   :cardinality 1572
+   :description "Public vars across all indexed namespaces. Gathered from JVM metadata, filtered by search/var-keys."
```
Comment (lambdasierra) [inline docs/entity-attribute-model.edn:100]: For now, this is good enough, but if we want to support multiple versions (of Clojure or any other library) in the future, we will have to change this representation.

In [Clojure terms](https://clojure.org/about/state), a Var is an **identity**, a name for something that can have **different** values over time.

Clojure makes it easy to talk about a `Var` as a single thing because Clojure has such a strong policy of preserving stability and backwards-compatibility. That means Clojure mostly only adds new Vars, and rarely changes the behavior of existing Vars.

However, even Clojure sometimes **adds new behavior** to existing Vars. For example, [new arities added to map and other sequence functions](https://github.com/clojure/clojure/commit/2a09172e0c3285ccdf79d1dc4d399d190678b670#diff-313df99869bda118262a387b322021d2cd9bea959fad3890cb78c6c37b448299R2563-R2571) with [Transducers](https://clojure.org/reference/transducers), which also updated the docstrings.

Other Clojure libraries **do not** have the same stability promises, even some of the old [clojure-contrib](https://github.com/clojure/clojure-contrib) libraries. Different versions of the same library may define different Vars, reusing the same namespace+name. In fact, there's nothing stopping two different libraries from defining the same Var, although this would be considered extremely poor taste and can probably be safely ignored.

So in the future, **if and when** we want to support multiple versions, we will have to distinguish between the inherent attributes of a Var (namespace and name) and the **state** of that Var (metadata and root binding) in a particular instance (a library "version"?).

If all this sounds complicated, that's because it is! Again, this is not something we need to deal with right now.

File: docs/entity-attribute-model.edn (hunk 14 above)
Comment (lambduhh) [reply docs/entity-attribute-model.edn:361]: relates to https://github.com/nubank/clojuredocs/pull/71#discussion_r3426357287 https://github.com/nubank/clojuredocs/pull/71#discussion_r3424006497
